sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/sycl_ext_oneapi_bindless_images.asciidoc

```diff
@@ -0,0 +1,541 @@
+  }
+}
+
+// Image descriptor - can use the same for all images
+sycl::ext::oneapi::image_descriptor desc(
+  {width, height}, image_channel_order::rgba, image_channel_type::fp32);
+
+// Allocate each image and save the handles
+std::vector<sycl::ext::oneapi::image_mem_handle> imgAllocations;
+for (int i = 0; i < numImages; i++) {
+  // Extension: returns the handle to the allocated memory
+  auto img_mem = sycl::ext::oneapi::allocate_image(ctxt, desc);
+  if (img_mem == nullptr) {
+    std::cout << "Error allocating image!" << std::endl;
+    return 1;
+  }
+  imgAllocations.push_back(img_mem);
+}
+
+// Copy over data to device for each image
+for (int i = 0; i < numImages; i++) {
+  // Extension: copy over data to device
+  q.ext_image_copy(imgAllocations[i], dataIn.data(), desc);
+}
+
+// Wait for copy operations to finish
+q.wait();
+
+// Create the images and return the handles
+std::vector<sycl::ext::oneapi::unsampled_image_handle> imgHandles;
+for (int i = 0; i < numImages; i++) {
+  // Extension: create the image and return the handle
+  sycl::ext::oneapi::unsampled_image_handle imgHandle =
+      sycl::ext::oneapi::create_image(ctxt, imgAllocations[i], desc);
+  imgHandles.push_back(imgHandle);
+}
+
+// Cuda stores data in column-major fashion
+// SYCL deals with indexing in row-major fashion
+// Reverse output buffer dimensions and access to convert
+// the cuda column-major data back to row-major
+buffer<float, 2> buf((float *)out.data(), range<2>{height, width});
+buffer<sycl::ext::oneapi::unsampled_image_handle, 1> imgHandlesBuf(
+    imgHandles.data(), range<1>{numImages});
+q.submit([&](handler &cgh) {
+  auto outAcc =
+      buf.get_access<access_mode::write>(cgh, range<2>{height, width});
+
+  auto imgHandleAcc =
+      imgHandlesBuf.get_access<access_mode::read>(cgh, range<1>{numImages});
+
+  cgh.parallel_for<image_addition>(
+      nd_range<2>{{width, height}, {width, height}}, [=](nd_item<2> it) {
+        size_t dim0 = it.get_local_id(0);
+        size_t dim1 = it.get_local_id(1);
+
+        // Sum each image by reading their handle
+        float sum = 0;
+        for (int i = 0; i < numImages; i++) {
+          // Extension: read image data from handle
+          sum += (sycl::ext::oneapi::read_image<float4>(
+              imgHandleAcc[i], int2(dim0, dim1)))[0];
+        }
+        outAcc[id<2>{dim1, dim0}] = sum;
+      });
+});
+
+// Using image handles requires manual synchronization
+q.wait_and_throw();
+
+// Cleanup
+for (int i = 0; i < numImages; i++) {
+  sycl::ext::oneapi::destroy_image_handle(ctxt, imgHandles[i]);
+  sycl::ext::oneapi::free_image(ctxt, imgAllocations[i]);
+}
+```
+
+== Implementation notes
+
+The current DPC++ prototype only implements the proposal for the CUDA backend,
+however we are actively exploring Level Zero with SPIR-V.
+We are looking at other backend as well in order to ensure the extension can 
+work across different backends.
+
+== Issues
```

Comment:
And another global comment.

I noted above that the image handle types need to define a copy constructor / operator in order to enable your examples.  One way to address this is to say that they have the "common reference semantics".  This does not give you a default constructor, though, so you would still need to specify that.

Defining `image_mem_handle` as an object with common reference semantics would have some nice benefits:

* The functions listed under "Getting image information from non-USM image memory" could all be member functions instead of free functions.
* The `image_mem_handle` object could store its context internally, which means that most of the functions operating on it would no longer need a context parameter.
* The `allocate_image` function would simply become a constructor for `image_mem_handle`.
* We would no longer need the `free_image` function because the destructor would automatically deallocate the memory (because it's common reference semantics, this would only happen when the last reference was dropped).
* The programming model would become RAII, which is generally considered a good C++ design principle.
* This type would be more consistent with other SYCL types, which also have common reference semantics.

Defining `unsampled_image_handle` and `sampled_image_handle` as common reference semantics would have similar nice benefits.